File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function(arr, k) {
  const n = arr.length;
  const dp = new Array(n + 1).fill(0);
  
  for (let i = 1; i <= n; i++) {
    let [currMax, max_sum] = [0, 0];
    for (let j = 1; j <= Math.min(k, i); j++) {
      currMax = Math.max(currMax, arr[i - j]);
      max_sum = Math.max(max_sum, dp[i - j] + currMax * j);
    }
    
    dp[i] = max_sum;
  }
  
  return dp[n];
};

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        n = len(arr)
        memo = [0] * (n + 1)
        for i in range(1, n + 1):
            curr_max, max_sum = 0, 0
            for j in range(1, min(k + 1, i + 1)):
                curr_max = max(curr_max, arr[i - j])
                max_sum = max(max_sum, memo[i - j] + curr_max * j)
            memo[i] = max_sum
        return memo[-1]